Framework/API/inc/MantidAPI/Algorithm.h

```diff
@@ -306,6 +307,34 @@ class MANTID_API_DLL Algorithm : public IAlgorithm,
   /// Virtual method - must be overridden by concrete algorithm
   virtual void exec() = 0;
 
+  virtual void execDistributed() { exec(); }
+  virtual void execMasterOnly();
+  virtual void execNonMaster(){};
```

Comment:
To clarify this a bit more: If a workspace has `StorageMode::MasterOnly`, the workspace still exists on *all* ranks, but there are not spectra on the non-master ranks. The reason for this is that then we can reuse existing Python reduction scripts without modification. It the workspace would really exist only on the master rank, each Python script would need to check for the rank everywhere and call algorithms only when the workspace is available.